### PR TITLE
Feature-flag gated tests to validate auto http2/gRPC detection with empty port

### PR DESCRIPTION
**Pull request by rafaeltello**
Creates new "emptyport" tests to validate protocol used when no port is specified. Currently only 1 of the tests runs all the time, the others are gated behind a new feature flag.

We originally intended to add a more complex support matrix in this change. However, while we can hack the `http.Transport.TLSNextProto` to force a http 1.1 connection, the way the dialer is implemented (by using a TCP connection to the ingress endpoint) makes it meaningless in this test. Conversely, using `http2.Transport` to force a http2-only connection requires us to create an equivalent version of https://github.com/knative/networking/blob/master/test/conformance/ingress/util.go#L902 using a slightly different signature.

We still want to do this, but will probably occur in parallel with the rest of development.

Work is being tracked by https://github.com/knative/serving/issues/4283

### CODE REVIEW

**Review by igorbelianski — APPROVED** [review]

File: test/conformance/ingress/emptyport.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (chaodaiG) [inline test/conformance/ingress/emptyport.go:2]:
```suggestion
Copyright 2020 The Knative Authors
```

File: test/conformance/ingress/emptyport.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+							ServicePort:      intstr.FromInt(port),
+						},
+					}},
+				}},
+			},
+		}},
+	})
+
+	ri := RuntimeRequest(ctx, t, client, "http://"+name+".example.com")
+	if ri == nil {
```
Comment (chaodaiG) [inline test/conformance/ingress/emptyport.go:93]: This test can succeed if ri is nil or some other condition, is it expected?

**Comment by tcnghia** [comment]
@rafaeltello can https://github.com/knative/networking/pull/184 be useful for this purpose? can you please check with @nak3? thanks!

File: test/conformance/ingress/emptyport.go (hunk 2 above)
Comment (rafaeltello) [reply test/conformance/ingress/emptyport.go:93]: This is essentially https://github.com/knative/networking/blob/master/test/conformance/ingress/basic.go#L60 with an empty port, so we intentionally want to keep the test as close as possible as that one.

**Comment by rafaeltello** [comment]
> @rafaeltello can #184 be useful for this purpose? can you please check with @nak3? thanks!

Yes!!! This is probably super useful, and I now need to merge with this file. Let me change that.

**Comment by rafaeltello** [comment]
> @rafaeltello can #184 be useful for this purpose? can you please check with @nak3? thanks!

I ended up keeping my original flag, as the feature is not even Alpha yet. I did add a TODO to move this into alphaTests once the feature is implemented :)

File: test/conformance/ingress/run.go [hunk 3]
```diff
@@ -93,6 +93,14 @@ func RunConformance(t *testing.T) {
 			t.Run(name, test)
 		}
 	}
+
+	// TODO(rafaeltc): Once the feature is partially implemented,
+	// move to alphaTests and remove flag.
```
Comment (tcnghia) [inline test/conformance/ingress/run.go:98]: Link to issue instead? Thanks

**Review by tcnghia — CHANGES_REQUESTED** [review]
/assign @nak3

File: test/conformance/ingress/emptyport.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"knative.dev/networking/pkg/apis/networking/v1alpha1"
+	"knative.dev/networking/test"
+	ping "knative.dev/networking/test/test_images/grpc-ping/proto"
+	"knative.dev/pkg/test/logstream"
+)
+
+// TestHTTP1AndEmptyPort verifies that an empty port name uses HTTP1. This should be the current behavior.
+func TestHTTP1AndEmptyPort(t *testing.T) {
```
Comment (nak3) [inline test/conformance/ingress/emptyport.go:34]: Can we add `t.Parallel()`?

File: test/conformance/ingress/emptyport.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	if want, got := 1, ri.Request.ProtoMajor; want != got {
+		t.Errorf("ProtoMajor = %d, wanted %d", got, want)
+	}
+}
+
+// TestHTTP2AndEmptyPort verifies that an empty port name uses HTTP2.
+// This is not the current behavior.
+func TestHTTP2AndEmptyPort(t *testing.T) {
```
Comment (nak3) [inline test/conformance/ingress/emptyport.go:69]: Same with above. Can we add `t.Parallel()`?

**Review by nak3 — COMMENTED** [review]
In addition these tests, should we add a traffic split test which has empty name port and non-empty name port mixed?

File: test/conformance/ingress/emptyport.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+			},
+		}},
+	})
+
+	ri := RuntimeRequest(ctx, t, client, "http://"+name+".example.com")
+	if ri == nil {
+		return
+	}
+
+	if want, got := 2, ri.Request.ProtoMajor; want != got {
```
Comment (nak3) [inline test/conformance/ingress/emptyport.go:97]: I'm sorry I may be stupid. But `TestHTTP1AndEmptyPort` and `TestHTTP2AndEmptyPort` are same tests and using same `runtime` image. How this test can get protocol version 2?

File: test/conformance/ingress/emptyport.go (hunk 4 above)
Comment (rafaeltello) [reply test/conformance/ingress/emptyport.go:34]: Will do! Sorry, I created these tests before that change went in, and missed copying it.

File: test/conformance/ingress/emptyport.go (hunk 6 above)
Comment (rafaeltello) [reply test/conformance/ingress/emptyport.go:99]: Oh! No, I'm the stupid one :( you're absolutely right; I had done some additional work here, but then I reverted most of it and didn't notice the two test cases were the same.
Let me rewrite this today.

**Comment by rafaeltello** [comment]
Thanks to @nak3 comment we discovered a more fundamental testability issue with this change. We need to test this in serving, not in networking, to actually exercise knative's eventual ability to autodetect http2/gRPC. These are k8s services.

I am closing this request, and sending equivalent ones for serving.
